### DISCUSSION TITLE
Extension API context should expose player source information

### DISCUSSION THREAD

**Discussion by rgov** [foxglove/community#293] [Q&A]
In particular, I would find it valuable to know whether my panel is getting data from a live system or is being played back from a rosbag/mcap file.

Use cases:

* Hide tele-op / interactive elements when playing back a recording
* Display an error if the Player does not have required [capabilities](https://github.com/foxglove/studio/blob/33033e2cb477ceecdfb8191d0e5ac02f621ab0b2/packages/studio-base/src/players/types.ts#L296)

Honestly maybe this request should just be "expose `PlayerCapabilities` through the panel context object".

**Comment by defunctzombie** [comment]
The way this is currently handled in the extension API is by having various methods on the context be undefined. So if the data source does not support publishing, the `advertise` and `publish` methods won't be available. If the data source does not support services, the `callService` method won't be available. etc.

This is separate from knowing if the connection is live or not tho. For file data source those two methods are not available iirc.

If this is too implicit then we could consider adding an explicit list of capabilities but today that is done as described above.

**Comment by daf** [comment]
We've written a small HTTP service that lives alongside our ROS setup and is able to interact with it.  We've also developed a custom Foxglove Studio panel that needs to interface with it - in nearly every case I can think of, the address to that HTTP service will mirror that of the active connection in Foxglove Studio.  It would be useful if this information is given via the `PanelExtensionContext` parameter that's provided to the custom panel.

Currently it's implemented via the normal panel configuration stuff, but with several different deployment environments, this doesn't scale well when using shared Layouts.  I'd love to just get the player connection info and do some string replacement (with config override).  Is this something folks would be ok with adding?

This is a slightly different use case than what's described above.

**Reply by defunctzombie** [reply]
Thanks for sharing additional context about your workflow. For now, our stance on this remains as described in this comment: https://github.com/foxglove/studio/pull/5672#issuecomment-1526794477
